File: ios/SkyCast/Core/DesignSystem/MetricVisuals.swift
import SwiftUI

/// How a reading draws itself.
///
/// Each case is a shape that suits a *kind* of reading rather than a specific metric, so the design
/// system decides how things look and the feature layer only says which kind it has.
enum MetricVisual: Equatable {
    /// Nothing but the number. For readings with no scale worth drawing.
    case plain
    /// A share of a known range, 0 to 1, on an arc. The middle of the tile stays free for the
    /// value, so the number remains the largest thing in it.
    case gauge(Double)
    /// A bearing in degrees, with its compass point already named. For wind.
    case compass(degrees: Double, cardinal: String)
}

/// An arc showing where a reading sits on its own scale.
///
/// 240° opening downwards, so the gap gives the eye a start and an end.
struct MetricGauge: View {
    let fraction: Double
    let colour: Color

    private var clamped: Double {
        min(max(fraction, 0), 1)
    }

    var body: some View {
        ZStack {
            arc(to: 1)
                .stroke(Color.primary.opacity(trackOpacity), style: stroke)
            arc(to: clamped)
                .stroke(colour, style: stroke)
        }
        .frame(width: diameter, height: diameter)
        // The value beside it is the fact; the arc restates it. Announcing both would be a stutter.
        .accessibilityHidden(true)
        // Length is geometry, so it animates, but only when the value genuinely changes.
        .animation(.smooth, value: clamped)
    }

    private func arc(to end: Double) -> Path {
        Path { path in
            let centre = CGPoint(x: diameter / 2, y: diameter / 2)
            path.addArc(
                center: centre,
                radius: diameter / 2 - lineWidth / 2,
                startAngle: .degrees(Self.startDegrees),
                endAngle: .degrees(Self.startDegrees + Self.sweepDegrees * end),
                clockwise: false
            )
        }
    }

    private var stroke: StrokeStyle {
        StrokeStyle(lineWidth: lineWidth, lineCap: .round)
    }

    /// 150° puts the opening at the bottom, symmetrical about vertical.
    private static let startDegrees: Double = 150
    private static let sweepDegrees: Double = 240

    private let diameter: CGFloat = 44
    private let lineWidth: CGFloat = 5
    private let trackOpacity: Double = 0.12
}

/// A compass rose with a needle on the bearing the wind is coming *from*.
///
/// Meteorological convention: 270° is a westerly, blowing from the west. The needle points at the
/// source, which is what "westerly" means.
struct WindCompass: View {
    let degrees: Double
    let colour: Color

    var body: some View {
        ZStack {
            Circle()
                .stroke(Color.primary.opacity(ringOpacity), lineWidth: 1)

            // North tick, so the rose has an orientation without four letters cluttering it.
            Rectangle()
                .fill(Color.primary.opacity(tickOpacity))
                .frame(width: 1, height: tickLength)
                .offset(y: -diameter / 2 + tickLength / 2)

            Capsule()
                .fill(colour)
                .frame(width: needleWidth, height: diameter * needleLengthRatio)
                .offset(y: -diameter * needleOffsetRatio)
                .rotationEffect(.degrees(degrees))
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
        .animation(.smooth, value: degrees)
    }

    private let diameter: CGFloat = 44
    private let ringOpacity: Double = 0.15
    private let tickOpacity: Double = 0.3
    private let tickLength: CGFloat = 5
    private let needleWidth: CGFloat = 3
    private let needleLengthRatio: CGFloat = 0.42
    private let needleOffsetRatio: CGFloat = 0.21
}

/// The sun's day, as an arc from sunrise to sunset with a marker at now.
///
/// The length of the day is the caption.
struct SunPathCard: View {
    /// 0 at sunrise, 1 at sunset. Outside that range the sun is down and the marker is hidden.
    let progress: Double
    let sunriseLabel: String
    let sunsetLabel: String
    let daylightLabel: String
    /// Spoken instead of the drawing.
    let announcement: String

    var body: some View {
        SkyPathCard(
            progress: progress,
            riseLabel: sunriseLabel,
            setLabel: sunsetLabel,
            centreLabel: daylightLabel,
            riseSymbol: "sunrise.fill",
            setSymbol: "sunset.fill",
            riseColour: WeatherPalette.sunrise,
            setColour: WeatherPalette.sunset,
            markerColour: WeatherPalette.sunrise,
            announcement: announcement
        )
    }
}

/// An arc from a rise to a set, with a marker at the body's current position.
///
/// Shared by the sun card and the moon card, which differ only in times, colours and symbols.
struct SkyPathCard: View {
    /// 0 at the rise, 1 at the set. Outside that range the body is below the horizon and the marker
    /// is hidden.
    let progress: Double
    let riseLabel: String
    let setLabel: String
    let centreLabel: String
    let riseSymbol: String
    let setSymbol: String
    let riseColour: Color
    let setColour: Color
    let markerColour: Color
    /// Spoken instead of the drawing.
    let announcement: String


    private var isUp: Bool {
        progress > 0 && progress < 1
    }

    var body: some View {
        VStack(spacing: Spacing.sm) {
            GeometryReader { proxy in
                let size = proxy.size
                ZStack(alignment: .topLeading) {
                    SkyArc()
                        .stroke(
                            Color.primary.opacity(trackOpacity),
                            style: StrokeStyle(lineWidth: 2, dash: [4, 4])
                        )

                    SkyArc()
                        .trim(from: 0, to: min(max(progress, 0), 1))
                        .stroke(
                            .linearGradient(
                                colors: [riseColour, setColour],
                                startPoint: .leading,
                                endPoint: .trailing
                            ),
                            style: StrokeStyle(lineWidth: 3, lineCap: .round)
                        )

                    if isUp {
                        Circle()
                            .fill(markerColour)
                            .frame(width: markerDiameter, height: markerDiameter)
                            .position(SkyArc.point(at: progress, in: size))
                    }
                }
            }
            .frame(height: arcHeight)

            HStack {
                label(riseLabel, systemImage: riseSymbol, colour: riseColour)
                Spacer()
                Text(centreLabel)
                    .font(.caption.weight(.medium))
                    .foregroundStyle(.secondary)
                Spacer()
                label(setLabel, systemImage: setSymbol, colour: setColour)
            }
        }
        .padding(Spacing.md)
        .frostedCard()
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(announcement)
    }

    private func label(_ text: String, systemImage: String, colour: Color) -> some View {
        Label {
            Text(text).font(.caption).monospacedDigit()
        } icon: {
            Image(systemName: systemImage)
                .font(.caption)
                .foregroundStyle(colour)
        }
    }

    private let arcHeight: CGFloat = 64
    private let markerDiameter: CGFloat = 10
    private let trackOpacity: Double = 0.15
}

/// A shallow arc across a tile, flat enough to read as a horizon rather than a rainbow.
///
/// Shared by the sun card and the moon card.
struct SkyArc: Shape {
    func path(in rect: CGRect) -> Path {
        Path { path in
            path.move(to: CGPoint(x: 0, y: rect.maxY))
            path.addQuadCurve(
                to: CGPoint(x: rect.maxX, y: rect.maxY),
                control: CGPoint(x: rect.midX, y: rect.minY - rect.height * Self.controlLift)
            )
        }
    }

    /// The point on the same curve at `progress`, for placing the marker.
    ///
    /// Evaluated from the quadratic directly. The parameter is not arc length, but for a curve this
    /// shallow the difference is under a pixel.
    static func point(at progress: Double, in size: CGSize) -> CGPoint {
        let along = min(max(progress, 0), 1)
        let start = CGPoint(x: 0, y: size.height)
        let end = CGPoint(x: size.width, y: size.height)
        let control = CGPoint(x: size.width / 2, y: -size.height * controlLift)
        let inverse = 1 - along
        return CGPoint(
            x: inverse * inverse * start.x + 2 * inverse * along * control.x + along * along * end.x,
            y: inverse * inverse * start.y + 2 * inverse * along * control.y + along * along * end.y
        )
    }

    /// Lifts the control point above the tile so the drawn curve peaks inside it.
    private static let controlLift: CGFloat = 0.6
}

#Preview("Gauge and compass") {
    HStack(spacing: Spacing.lg) {
        MetricGauge(fraction: 0.78, colour: WeatherPalette.humidity)
        MetricGauge(fraction: 0.2, colour: WeatherPalette.pressure)
        WindCompass(degrees: 270, colour: WeatherPalette.wind)
    }
    .padding()
}

#Preview("Sun path") {
    SunPathCard(
        progress: 0.62,
        sunriseLabel: "05:50",
        sunsetLabel: "20:18",
        daylightLabel: "14h 28m",
        announcement: "Sunrise 05:50, sunset 20:18, 14 hours 28 minutes of daylight"
    )
    .padding()
    .background(Color.skyBackground)
}

File: android/core/designsystem/src/main/kotlin/com/nauhaan/skycast/core/designsystem/component/MetricVisuals.kt
package com.nauhaan.skycast.core.designsystem.component

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material.icons.filled.WbTwilight
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.compositeOver
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.nauhaan.skycast.core.designsystem.theme.LocalWeatherTint
import com.nauhaan.skycast.core.designsystem.theme.SkyCastTheme
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.core.designsystem.theme.weatherPalette

/**
 * How a reading draws itself.
 *
 * Each case is a shape that suits a *kind* of reading rather than a specific metric, so the design
 * system decides how things look and the feature layer only says which kind it has.
 *
 * Mirrors `MetricVisual` on iOS.
 */
@Immutable
sealed interface MetricVisual {
    /** Nothing but the number, for readings with no scale worth drawing. */
    data object Plain : MetricVisual

    /**
     * A share of a known range, 0–1, on an arc.
     *
     * Reads as "how full" at a glance, which a bar also does, but an arc leaves the middle of the
     * tile free, so the number stays the largest thing in it.
     */
    data class Gauge(val fraction: Float) : MetricVisual

    /** A bearing in degrees, with its compass point already named. For wind. */
    data class Compass(val degrees: Float, val cardinal: String) : MetricVisual
}

/**
 * An arc showing where a reading sits on its own scale.
 *
 * 240° rather than a full circle, opening downwards: a closed ring reads as a progress spinner, and
 * the gap gives the eye a start and an end so "nearly full" is unambiguous.
 */
@Composable
fun MetricGauge(fraction: Float, colour: Color, modifier: Modifier = Modifier) {
    val animated by animateFloatAsState(
        targetValue = fraction.coerceIn(0f, 1f),
        // Effects, not spatial: an arc that overshoots shows a value the reading never had.
        animationSpec = MaterialTheme.motionScheme.defaultEffectsSpec(),
        label = "metricGauge",
    )
    val track = weatherPalette.metricTrack

    Canvas(modifier = modifier.size(GaugeDiameter)) {
        val stroke = Stroke(width = GaugeStrokePx, cap = StrokeCap.Round)
        val inset = GaugeStrokePx / 2
        val arcSize = Size(size.width - GaugeStrokePx, size.height - GaugeStrokePx)
        drawArc(
            color = track,
            startAngle = GAUGE_START_DEGREES,
            sweepAngle = GAUGE_SWEEP_DEGREES,
            useCenter = false,
            topLeft = Offset(inset, inset),
            size = arcSize,
            style = stroke,
        )
        drawArc(
            color = colour,
            startAngle = GAUGE_START_DEGREES,
            sweepAngle = GAUGE_SWEEP_DEGREES * animated,
            useCenter = false,
            topLeft = Offset(inset, inset),
            size = arcSize,
            style = stroke,
        )
    }
}

/**
 * A compass rose with a needle on the bearing the wind is coming *from*.
 *
 * Meteorological convention: 270° is a westerly, blowing from the west. The needle points at the
 * source, which is what "westerly" means and what every aviation and marine chart shows, drawing it
 * as an arrow in the direction of travel would silently invert the reading for anyone who knows the
 * convention.
 */
@Composable
fun WindCompass(degrees: Float, colour: Color, modifier: Modifier = Modifier) {
    val animated by animateFloatAsState(
        targetValue = degrees,
        animationSpec = MaterialTheme.motionScheme.defaultSpatialSpec(),
        label = "windCompass",
    )
    val ring = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = RING_ALPHA)

    Canvas(
        modifier = modifier
            .size(GaugeDiameter)
            // Rotating the layer rather than the maths keeps the needle a straight line.
            .graphicsLayer { rotationZ = animated },
    ) {
        val radius = size.minDimension / 2
        val centre = Offset(size.width / 2, size.height / 2)
        drawCircle(color = ring, radius = radius - 1, style = Stroke(width = 1f))
        drawLine(
            color = colour,
            start = Offset(centre.x, centre.y - radius * NEEDLE_INNER),
            end = Offset(centre.x, centre.y - radius * NEEDLE_OUTER),
            strokeWidth = NeedleWidthPx,
            cap = StrokeCap.Round,
        )
    }
}

/** The sun-path card's content, already formatted by the feature layer. */
@Immutable
data class SunPathReading(
    val progress: Float,
    val sunriseLabel: String,
    val sunsetLabel: String,
    val daylightLabel: String,
    val contentDescription: String,
)

/**
 * The sun's day, as an arc from sunrise to sunset with a marker at now.
 *
 * This replaces two tiles that each showed a bare time. "05:50" and "20:18" are facts the reader has
 * to do arithmetic on before they mean anything; an arc with the sun three quarters along it answers
 * "how much daylight is left?" without any. The length of the day comes along as the caption, so a
 * third tile disappears too.
 */
@Composable
fun SunPathCard(reading: SunPathReading, modifier: Modifier = Modifier) {
    SkyPathCard(
        reading = SkyPathReading(
            progress = reading.progress,
            riseLabel = reading.sunriseLabel,
            setLabel = reading.sunsetLabel,
            centreLabel = reading.daylightLabel,
            contentDescription = reading.contentDescription,
        ),
        riseIcon = Icons.Filled.WbSunny,
        setIcon = Icons.Filled.WbTwilight,
        riseColour = weatherPalette.sunrise,
        setColour = weatherPalette.sunset,
        modifier = modifier,
    )
}

/** A rise-to-set card's content, already formatted by the feature layer. */
@Immutable
data class SkyPathReading(
    /** 0 at the rise, 1 at the set. Outside that range the body is below the horizon. */
    val progress: Float,
    val riseLabel: String,
    val setLabel: String,
    val centreLabel: String,
    val contentDescription: String,
)

/**
 * An arc from a rise to a set, with a marker at the body's current position.
 *
 * Generalised out of [SunPathCard] when the Moon tab needed the same drawing with different times,
 * colours and icons. Everything that makes it *readable*, the dashed remainder, the tinted
 * container, the single spoken sentence, is therefore defined once.
 */
@Composable
fun SkyPathCard(
    reading: SkyPathReading,
    riseIcon: ImageVector,
    setIcon: ImageVector,
    riseColour: Color,
    setColour: Color,
    modifier: Modifier = Modifier,
    markerColour: Color = riseColour,
) {
    val tint = LocalWeatherTint.current
    val base = MaterialTheme.colorScheme.surfaceContainerHigh
    val container = if (tint == null) base else tint.copy(alpha = TINT_ALPHA).compositeOver(base)
    val track = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = TRACK_ALPHA)
    val progress = reading.progress.coerceIn(0f, 1f)
    val isUp = reading.progress > 0f && reading.progress < 1f

    Card(
        modifier = modifier
            .fillMaxWidth()
            .clearAndSetSemantics { contentDescription = reading.contentDescription },
        colors = CardDefaults.cardColors(containerColor = container),
    ) {
        Column(modifier = Modifier.padding(Spacing.md)) {
            Canvas(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(ArcHeight),
            ) {
                val path = sunArc(size)
                drawPath(
                    path = path,
                    color = track,
                    style = Stroke(
                        width = 2f,
                        pathEffect = PathEffect.dashPathEffect(floatArrayOf(10f, 10f)),
                    ),
                )
                // The travelled part is drawn as its own shorter arc rather than by trimming the
                // path: Compose has no `trim`, and re-deriving the curve for a partial sweep is
                // exact where clipping the full one would flatten its end.
                drawPath(
                    path = sunArc(size, until = progress),
                    brush = Brush.horizontalGradient(listOf(riseColour, setColour)),
                    style = Stroke(width = 4f, cap = StrokeCap.Round),
                )
                if (isUp) {
                    drawCircle(
                        color = markerColour,
                        radius = MarkerRadiusPx,
                        center = sunArcPoint(progress, size),
                    )
                }
            }

            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = Spacing.sm),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                TimeLabel(reading.riseLabel, riseIcon, riseColour)
                Text(
                    text = reading.centreLabel,
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                TimeLabel(reading.setLabel, setIcon, setColour)
            }
        }
    }
}

/** Rise and set icons for a body with no dawn or dusk of its own. */
val MoonRiseIcon: ImageVector get() = Icons.Filled.ArrowUpward

/** @see MoonRiseIcon */
val MoonSetIcon: ImageVector get() = Icons.Filled.ArrowDownward

@Composable
private fun TimeLabel(text: String, icon: androidx.compose.ui.graphics.vector.ImageVector, tint: Color) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = tint,
            modifier = Modifier.size(LabelIconSize),
        )
        Text(
            text = text,
            style = MaterialTheme.typography.labelLarge,
            modifier = Modifier.padding(start = Spacing.xs),
        )
    }
}

/**
 * A shallow quadratic arc across the box, flat enough to read as a horizon rather than a rainbow.
 *
 * [until] draws only the first part of it, by evaluating the curve rather than clipping.
 */
private fun sunArc(size: Size, until: Float = 1f): Path = Path().apply {
    val start = sunArcPoint(0f, size)
    moveTo(start.x, start.y)
    val fraction = until.coerceIn(0f, 1f)
    val steps = (SUN_ARC_STEPS * fraction).toInt().coerceAtLeast(1)
    for (step in 1..steps) {
        // `step / steps`, not `step / SUN_ARC_STEPS`. The latter scales by the fraction twice, so
        // the drawn arc stopped at progress², visibly short of the sun marker, which is evaluated
        // from the same curve at the real progress. Caught by looking at it.
        val point = sunArcPoint(fraction * step / steps, size)
        lineTo(point.x, point.y)
    }
}

/** The point on that curve at [progress], for placing the marker and for drawing partial arcs. */
private fun sunArcPoint(progress: Float, size: Size): Offset {
    val t = progress.coerceIn(0f, 1f)
    val inverse = 1 - t
    val controlY = -size.height * SUN_ARC_LIFT
    return Offset(
        x = 2 * inverse * t * (size.width / 2) + t * t * size.width,
        y = inverse * inverse * size.height + 2 * inverse * t * controlY + t * t * size.height,
    )
}

private val GaugeDiameter = 44.dp
private val ArcHeight = 64.dp
private val LabelIconSize = 16.dp
private const val GaugeStrokePx = 12f
private const val NeedleWidthPx = 7f
private const val MarkerRadiusPx = 12f

/** 150° puts the gauge's opening at the bottom, symmetrical about vertical. */
private const val GAUGE_START_DEGREES = 150f
private const val GAUGE_SWEEP_DEGREES = 240f
private const val NEEDLE_INNER = 0.1f
private const val NEEDLE_OUTER = 0.82f
private const val RING_ALPHA = 0.2f
private const val TRACK_ALPHA = 0.25f
private const val TINT_ALPHA = 0.45f

/** Lifts the control point above the box so the drawn curve peaks inside it. */
private const val SUN_ARC_LIFT = 0.6f

/** Enough segments that the polyline reads as a curve at this size. */
private const val SUN_ARC_STEPS = 48

@Preview(showBackground = true)
@Composable
private fun MetricVisualsPreview() {
    SkyCastTheme {
        Column(modifier = Modifier.padding(Spacing.md)) {
            Row(horizontalArrangement = Arrangement.spacedBy(Spacing.lg)) {
                MetricGauge(fraction = 0.78f, colour = weatherPalette.humidity)
                MetricGauge(fraction = 0.2f, colour = weatherPalette.pressure)
                WindCompass(degrees = 270f, colour = weatherPalette.wind)
            }
            SunPathCard(
                reading = SunPathReading(
                    progress = 0.62f,
                    sunriseLabel = "05:50",
                    sunsetLabel = "20:18",
                    daylightLabel = "14h 28m",
                    contentDescription = "Sunrise 05:50, sunset 20:18",
                ),
                modifier = Modifier.padding(top = Spacing.md),
            )
        }
    }
}
